Man/basic-stats.1
.TH basic-stats 1
.SH NAME    \" Section header
.PP

basic-stats - Command-line statistics made trivial

\" Convention:
\" Underline anything that is typed verbatim - commands, etc.
.SH SYNOPSIS
.PP
.nf 
.na
basic-stats z-score x mean stddev
basic-stats z-cdf z-score [mean stddev] (defaults: 0 1)
basic-stats t-score x-bar expected-mean stddev n

basic-stats [--verbose] [--delim string] \\
	    function1 [param1] --row|--col N \\
	    [function2 ...]
.ad
.fi

.SH OPTIONS
.TP
\fB\-\-verbose
Echo input data during processing

.TP
\fB\-\-delim \fIstring\fR
Use string as a set of delimiters instead of the default " \\t"

.TP
\fB\-\-row \fIN\fR
Compute statistics for row N (1-based) of the input

.TP
\fB\-\-col \fIN\fR
Compute statistics for column N (1-based) of the input

.TP
\fBmean\fI --row|--col N\fR
Compute the mean (average) for row/column N

.TP
\fBquantile partitions\fI --row|--col N\fR
Compute the low, high, and quantiles for row/column N

.TP
\fBmedian\fI --row|--col N\fR
Compute the low, high, and median for row/column N.  Identical to
quantile 2 --row|--col N.

.TP
\fBquartile\fI --row|--col N\fR
Identical to quantile 4 --row|--col N.

.TP
\fBpop-var\fI --row|--col N\fR
Report population variance

.TP
\fBpop-stddev\fI --row|--col N\fR
Report population standard deviation

.TP
\fBpop-z-scores\fI --row|--col N\fR
Report population z-scores

.TP
\fBsample-var\fI --row|--col N\fR
Report sample variance

.TP
\fBsample-stddev\fI --row|--col N\fR
Report sample standard deviation

.TP
\fBsample-z-scores\fI --row|--col N\fR
Report sample z-scores.  Should only be used for sample sizes of 30 or more.
Otherwise, use sample-t-score.

.TP
\fBsample-stderr\fI --row|--col N\fR
Report sample standard error (sigma-x-bar = stddev / sqrt(n))

.TP
\fBsample-t-score\fI --row|--col N\fR
Report t-score for sample data.

.TP
\fBmode\fI --row|--col N\fR
Report mode

.TP
\fBrange\fI --row|--col N\fR
Report range

.TP
\fBiq-range\fI --row|--col N\fR
Report interquartile range

.SH "PURPOSE"

.B Basic-stats
is a program for quickly computing statistics on tabular data from the
command-line.  It elimintes the nuisance of importing data into a spreadsheet
or writing custom scripts in order to gather basic statistics.

.SH "DESCRIPTION"

.B Basic-stats
computes common statistics such as mean, median, etc. on rows and/or columns
of tabular data.  It can compute multiple functions on multiple rows and/or
columns in a single pass through the input stream.

The default delimiter is either a single TAB or multiple space characters. 
Alternate delimiters, such as a comma, can be specified using, e.g.
--delim ','.

Quantiles are computed using a generalized interpolation formula (method 4
from the Wikipedia page on quartiles).

The output format is designed to be both human-readable and easily parsed
by command-line tools and scripts.

.B Basic-stats is written entirely in C for maximal performance.  Memory use
is trivial for many functions, but some computations such as quantiles
which involve sorting the data require the use of arrays.

.SH EXAMPLES
.nf
.na
> cat sample.txt    
1    5  21
2   12  22
3   13  23
4   14  24
5   15  25
6   16  26
7   17  27
8   18  28
9   19  29
10  20  30
> basic-stats mean --col 1 median --row 2 quantile 4 --col 3 < sample.txt
Col 1 mean           5.500000
Row 2 low            2.000000
Row 2 quantile(0.50) 12.000000
Row 2 high           22.000000
Col 3 low            21.000000
Col 3 quantile(0.25) 22.750000
Col 3 quantile(0.50) 25.500000
Col 3 quantile(0.75) 28.250000
Col 3 high           30.000000
.ad
.fi

.SH "SEE ALSO"
awk(1), R(1)

.SH BUGS
Please report bugs to the author and send patches in unified diff format.
(man diff for more information)

.SH AUTHOR
.nf
.na
J. Bacon


basic-stats.c
/***************************************************************************
 *  Description:
 *      Command-line tool for quick, basic statistics.  Basic Unix tools
 *      don't facilitate stats very well and going to R for every little
 *      thing can really cramp your workflow, especially when using non-R
 *      tools for most of the work.
 *
 *  History: 
 *  Date        Name        Modification
 *  2020-03-24  Jason Bacon Begin
 ***************************************************************************/

#include <stdio.h>
#include <sysexits.h>
#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <xtend/dsv.h>
#include <xtend/mem.h>
#include "basic-stats.h"
#include "statsf-list.h"

// http://makemeanalyst.com/basic-statistics-for-data-analysis/

int     main(int argc, char *argv[])

{
    char    *delims = " \t";
    int     c,
	    flags = SFL_FLAG_NONE;
    statsf_list_t flist;

    if ( (argc == 2) && (strcmp(argv[1], "z-table") == 0) )
	return print_z_table();
    else if ( ((argc == 3) || (argc == 5)) && (strcmp(argv[1], "normal-cdf") == 0) )
	return print_normal_cdf(argc, argv);
    else if ( (argc == 5) && (strcmp(argv[1], "z-score") == 0) )
	return print_z_score(argc, argv);
    else if ( (argc == 5) && (strcmp(argv[1], "t-table") == 0) )
	return print_t_table();
    else if ( (argc == 6) && (strcmp(argv[1], "t-score") == 0) )
	return print_t_score(argc, argv);
    else if ( (argc == 4) && (strcmp(argv[1], "t-cdf") == 0) )
	return print_t_cdf(argc, argv);
    else if ( argc < 4 )
	usage(argv);
    
    statsf_list_init(&flist);
    for (c = 1; c < argc; ++c)
    {
	if ( strcmp(argv[c], "--delims") == 0 )
	{
	    if ( ++c < argc )
		delims = argv[c];
	    else
		usage(argv);
	}
	else if ( strcmp(argv[c],"--help") == 0 )
	    usage(argv);
	else if ( strcmp(argv[c],"--verbose") == 0 )
	    flags |= SFL_FLAG_VERBOSE;
	else if ( strcmp(argv[c],"quantile") == 0 )
	    statsf_list_add_func(&flist, STATSF_QUANTILE, &c, argc, argv);
	else if ( strcmp(argv[c],"median") == 0 )
	    statsf_list_add_func(&flist, STATSF_MEDIAN, &c, argc, argv);
	else if ( strcmp(argv[c],"quartile") == 0 )
	    statsf_list_add_func(&flist, STATSF_QUARTILE, &c, argc, argv);
	else if ( strcmp(argv[c],"mean") == 0 )
	    statsf_list_add_func(&flist, STATSF_MEAN, &c, argc, argv);
	else if ( strcmp(argv[c],"pop-var") == 0 )
	    statsf_list_add_func(&flist, STATSF_POP_VAR, &c, argc, argv);
	else if ( strcmp(argv[c],"sample-var") == 0 )
	    statsf_list_add_func(&flist, STATSF_SAMPLE_VAR, &c, argc, argv);
	else if ( strcmp(argv[c],"pop-stddev") == 0 )
	    statsf_list_add_func(&flist, STATSF_POP_STDDEV, &c, argc, argv);
	else if ( strcmp(argv[c],"pop-z-scores") == 0 )
	    statsf_list_add_func(&flist, STATSF_POP_Z_SCORES, &c, argc, argv);
	else if ( strcmp(argv[c],"sample-stddev") == 0 )
	    statsf_list_add_func(&flist, STATSF_SAMPLE_STDDEV, &c, argc, argv);
	else if ( strcmp(argv[c],"sample-stderr") == 0 )
	    statsf_list_add_func(&flist, STATSF_SAMPLE_STDERR, &c, argc, argv);
	else if ( strcmp(argv[c],"sample-z-scores") == 0 )
	    statsf_list_add_func(&flist, STATSF_SAMPLE_Z_SCORES, &c, argc, argv);
	else if ( strcmp(argv[c],"sample-t-score") == 0 )
	    statsf_list_add_func(&flist, STATSF_T_SCORE, &c, argc, argv);
	else if ( strcmp(argv[c],"mode") == 0 )
	    statsf_list_add_func(&flist, STATSF_MODE, &c, argc, argv);
	else if ( strcmp(argv[c],"range") == 0 )
	    statsf_list_add_func(&flist, STATSF_RANGE, &c, argc, argv);
	else if ( strcmp(argv[c],"iq-range") == 0 )
	    statsf_list_add_func(&flist, STATSF_IQ_RANGE, &c, argc, argv);
	else if ( strcmp(argv[c],"box-plot") == 0 )
	    statsf_list_add_func(&flist, STATSF_BOX_PLOT, &c, argc, argv);
	else if ( strcmp(argv[c],"chi-sq-fit") == 0 )
	    statsf_list_add_func(&flist, STATSF_CHI_FIT, &c, argc, argv);
	else if ( strcmp(argv[c],"chi-sq-ind") == 0 )
	    statsf_list_add_func(&flist, STATSF_CHI_IND, &c, argc, argv);
	else
	    usage(argv);
    }
    
    return statsf_list_process_stream(&flist, stdin, delims, flags);
}


void    usage(char *argv[])

{
    fprintf(stderr, "\nSimple functions:\n");
    fprintf(stderr, "   %s z-table\n", argv[0]);
    fprintf(stderr, "   %s z-score x mean stddev\n", argv[0]);
    fprintf(stderr, "   %s normal-cdf x [mean stddev] (defaults: 0 1)\n\n", argv[0]);
    fprintf(stderr, "   %s t-score x-bar expected-mean stddev n\n", argv[0]);
    fprintf(stderr, "   %s t-cdf x-bar n\n", argv[0]);
    
    fprintf(stderr, "Tabular data:\n");
    fprintf(stderr, "   %s [--verbose] [--delim 'string'] \\\n"
		    "       function1 [param1] --row|--col N1 \\\n"
		    "       [function2 [param2] --row|--col N2 ...] \\\n"
		    "       < input-file\n", argv[0]);
    fprintf(stderr, "\nFor tabular data at least one of the following functions is required:\n\n");
    fprintf(stderr, "  mean\n");
    fprintf(stderr, "  quantile N (N = number of divisions)\n");
    fprintf(stderr, "  median (same as quantile 2)\n");
    fprintf(stderr, "  quartile (same as quantile 4)\n");
    fprintf(stderr, "  pop-var\n");
    fprintf(stderr, "  pop-stddev\n");
    fprintf(stderr, "  pop-z-scores\n");
    fprintf(stderr, "  sample-var\n");
    fprintf(stderr, "  sample-stddev\n");
    fprintf(stderr, "  sample-z-scores\n");
    fprintf(stderr, "  sample-stderr\n");
    fprintf(stderr, "  sample-t-score expected-mean\n");
    fprintf(stderr, "  mode\n");
    fprintf(stderr, "  range\n");
    fprintf(stderr, "  iq-range\n");
    fprintf(stderr, "  box-plot\n");
    fprintf(stderr, "\nDefault delimiter is TAB.  The delimiter string may be multiple characters.\n");
    fprintf(stderr, "Multiple consecutive spaces are treated as a single delimiter.\n");
    fprintf(stderr, "\nExample: basic-stats sample-stddev --col 9 pop-var --col 10 < sample.tsv\n");
    fprintf(stderr, "Example: echo '3 6 1 8 2' | basic-stats mean --row 1\n\n");
    exit(EX_USAGE);
}


int     print_z_table(void)

{
    double  row, col;

    printf("      ");
    for (col = 0.00; col < 0.099; col += 0.01)
	printf("%6.2f ", col);
    putchar('\n');
    for (row = -3.0; row < 3.1; row += 0.1)
    {
	printf("%4.1f  ", row);
	for (col = 0.00; col < 0.099; col += 0.01)
	    printf("%.4f ", normal_cdf(row + (row < 0 ? -col : col), 0.0, 1.0));
	putchar('\n');
    }
    return EX_OK;
}


int     print_z_score(int argc, char *argv[])

{
    double  x, mean, stddev, score, cdf;
    char    *end;
    
    x = strtod(argv[2], &end);
    mean = strtod(argv[3], &end);
    stddev = strtod(argv[4], &end);
    score = z_score(x, mean, stddev);
    printf("z-score = %f\n", score);
    printf("P(x < %f) = %f\n", x, cdf = normal_cdf(x, mean, stddev));
    printf("1 - P = %f\n", 1.0 - cdf);
    return EX_OK;
}


int     print_normal_cdf(int argc, char *argv[])

{
    double  x, mean, stddev, cdf;
    char    *end;
    
    if ( argc == 3 )
    {
	x = strtod(argv[2], &end);
	if ( *end != '\0' )
	{
	    fprintf(stderr, "Invalid z-score: %s\n", argv[1]);
	    usage(argv);
	}
	printf("P(x < %f) = %f\n", x, cdf = normal_cdf(x, 0.0, 1.0));
	printf("1 - P = %f\n", 1.0 - cdf);
    }
    else if ( argc == 5 )
    {
	x = strtod(argv[2], &end);
	mean = strtod(argv[3], &end);
	stddev = strtod(argv[4], &end);
	printf("P(x < %f) = %f\n", x, cdf = normal_cdf(x, mean, stddev));
	printf("1 - P = %f\n", 1.0 - cdf);
    }
    else
	usage(argv);
    return EX_OK;
}


int     print_t_table(void)

{
    fprintf(stderr, "Not yet iumplemented.\n");
    return EX_OK;
}


int     print_t_score(int argc, char *argv[])

{
    double      x_bar, expected_mean, stddev, score, cdf;
    unsigned    n;
    char        *end;
    
    x_bar = strtod(argv[2], &end);
    expected_mean = strtod(argv[3], &end);
    stddev = strtod(argv[4], &end);
    n = strtoul(argv[5], &end, 10);
    score = t_score(x_bar, expected_mean, stddev, n);
    printf("t(%u) = %f\n", n - 1, score);
    printf("P(t(%u) < %f) = %f\n", n - 1, score, cdf = t_cdf(score, n));
    printf("1 - P = %f\n", 1.0 - cdf);
    return EX_OK;
}


int     print_t_cdf(int argc, char *argv[])

{
    double      x_bar, cdf;
    unsigned    n;
    char        *end;
    
    x_bar = strtod(argv[2], &end);
    n = strtoul(argv[3], &end, 10);
    printf("P(x_bar < %f) = %f\n", x_bar, cdf = t_cdf(x_bar, n));
    printf("1 - P = %f\n", 1.0 - cdf);
    
    return EX_OK;
}

